beacon_node/store/src/hot_cold_store.rs

```diff
@@ -1463,20 +1480,43 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
         state: &BeaconState<E>,
         ops: &mut Vec<KeyValueStoreOp>,
     ) -> Result<(), Error> {
-        // Put the state in the cache.
-        let block_root = state.get_latest_block_root(*state_root);
-
         // Avoid storing states in the database if they already exist in the state cache.
         // The exception to this is the finalized state, which must exist in the cache before it
         // is stored on disk.
-        if let PutStateOutcome::Duplicate =
-            self.state_cache
-                .lock()
-                .put_state(*state_root, block_root, state)?
-        {
+        match self.state_cache.lock().put_state(
+            *state_root,
+            state.get_latest_block_root(*state_root),
+            state,
+        )? {
+            PutStateOutcome::New(deleted_states) => {
+                debug!(
+                    self.log,
+                    "Cached state";
+                    "location" => "store_hot_state",
+                    "deleted_states" => ?deleted_states,
+                    "state_root" => ?state_root,
+                    "slot" => state.slot(),
+                );
+            }
+            PutStateOutcome::Duplicate => {
+                debug!(
+                    self.log,
+                    "State already exists in state cache";
+                    "slot" => state.slot(),
+                    "state_root" => ?state_root
+                );
+                // Write the state anyway. Testing of this branch shows that states may exist in the
+                // cache but not be yet ever stored in the DB
+            }
+            PutStateOutcome::Finalized => {} // ignore
+        }
+
+        // TODO(hdiff): is this optimization necessary? Computing diffs is expensive so we may want
+        // to avoid it.
+        if self.load_hot_state_summary(state_root)?.is_some() {
```

Comment:
An `TODO(hdiff)` sneaked in. This check relies on the assumption that states are only ever deleted on finalization pruning. Otherwise we need a lock. I'm not sure if this check is same in this branch.

Reply:
I've deleted this check, as I want to keep this PR minimal and `release-v7.0.0` is working OK wrt temporary states/etc AFAICT. The bugs we saw on tree-states hot were caused by partially removing temporary states.